psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -888,24 +975,437 @@ void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::v
+        auto printer = PsiOutStream("bench.dat", mode);
+        size_t ntri = nshell * (nshell + 1L) / 2L;
+        size_t possible_shells = ntri * (ntri + 1L) / 2L;
+        printer.Printf("Computed %20zu Shell Quartets out of %20zu, (%11.3E ratio)\n", computed_shells,
+                        possible_shells, computed_shells / (double)possible_shells);
+    }
+
+    timer_off("build_JK_matrices()");
+}
+
+// To follow this code, compare with figure 1 of DOI: 10.1063/1.476741
+void DirectJK::build_linK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, const std::vector<SharedMatrix>& D,
+                  std::vector<SharedMatrix>& K) {
+
+    if (!lr_symmetric_) {
+        throw PSIEXCEPTION("Non-symmetric K matrix builds are currently not supported in the LinK algorithm.");
+    }
+
+    timer_on("build_linK()");
+
+    // ==> Prep Auxiliary Quantities <== //
+
+    // => Zeroing <= //
+    for (auto& Kmat : K) {
+        Kmat->zero();
+    }
+
+    // => Sizing <= //
+    int nshell = primary_->nshell();
+    int nbf = primary_->nbf();
+    int nthread = df_ints_num_threads_;
+
+    // => Atom Blocking <= //
+    std::vector<int> shell_endpoints_for_atom;
+    std::vector<int> basis_endpoints_for_shell;
+
+    int atomic_ind = -1;
+    for (int P = 0; P < nshell; P++) {
+        if (primary_->shell(P).ncenter() > atomic_ind) {
+            shell_endpoints_for_atom.push_back(P);
+            atomic_ind++;
+        }
+        basis_endpoints_for_shell.push_back(primary_->shell_to_basis_function(P));
+    }
+    shell_endpoints_for_atom.push_back(nshell);
+    basis_endpoints_for_shell.push_back(nbf);
+
+    size_t natom = shell_endpoints_for_atom.size() - 1;
+
+    size_t max_functions_per_atom = 0L;
+    for (size_t atom = 0; atom < natom; atom++) {
+        size_t size = 0L;
+        for (int P = shell_endpoints_for_atom[atom]; P < shell_endpoints_for_atom[atom + 1]; P++) {
+            size += primary_->shell(P).nfunction();
+        }
+        max_functions_per_atom = std::max(max_functions_per_atom, size);
+    }
+
+    if (debug_) {
+        outfile->Printf("  ==> LinK: Atom Blocking <==\n\n");
+        for (size_t atom = 0; atom < natom; atom++) {
+            outfile->Printf("  Atom: %3d, Atom Start: %4d, Atom End: %4d\n", atom, shell_endpoints_for_atom[atom],
+                            shell_endpoints_for_atom[atom + 1]);
+            for (int P = shell_endpoints_for_atom[atom]; P < shell_endpoints_for_atom[atom + 1]; P++) {
+                int size = primary_->shell(P).nfunction();
+                int off = primary_->shell(P).function_index();
+                int off2 = basis_endpoints_for_shell[P];
+                outfile->Printf("    Shell: %4d, Size: %4d, Offset: %4d, Offset2: %4d\n", P, size, off,
+                                off2);
+            }
+        }
+        outfile->Printf("\n");
+    }
+
+    // ==> Prep Atom Pairs (Better for parallel performance than shell pairs) <== //
```

Comment:
That was based upon limited evidence when I was implementing LinK. Nonetheless, I removed the comment

Reply:
Removing the comment entirely is a mistake. When you make non-obvious design decisions, you document why you did them. Always always always. If your justification is "based upon limited evidence," say so explicitly. Don't hide it.

Reply:
"Better performance than shell pairs" just sounds like the wrong OpenMP scheduler was used...

Reply:
I agree with Susi; I'd prefer no comment over a wrong or misleading comment.

I read through the conversation, and it appears that Andy is screening pairs of atoms because he borrowed the DirectJK loop structure (which screens over pairs of atoms), and doing otherwise would require some refactoring:
https://github.com/psi4/psi4/pull/2359#pullrequestreview-830279803
https://github.com/psi4/psi4/pull/2359#issuecomment-992619129

There's only a brief mention of an attempted shell-based screening, but no code or timings are provided. It's not clear that a serious effort was made to compare these two schemes.
https://github.com/psi4/psi4/pull/2359#issuecomment-1003741024

To reiterate what previous reviewers have said, keeping the atom-pair blocking is perfectly fine, even if it's not optimal. In fact, I'd prefer no significant changes in this PR given the already difficult-to-follow history. What we don't want are misleading justifications in the code as to why that choice was made.

Reply:
"I'd prefer no comment over a wrong or misleading comment." is a false dichotomy. There should be a comment documenting why the decision was made and that it's open to change, following more dedicated investigation of parallelization.